src/base/kaldi-error.h

```diff
@@ -42,117 +43,139 @@ namespace kaldi {
 
 
 /***** KALDI LOGGING *****/
 
 /// Log message severity and source location info.
 struct LogMessageEnvelope {
+  /// Message severity. In addition to these levels, positive values (1 to 6)
+  /// specify verbose logging level. Verbose messages are produced only when
+  /// SetVerboseLevel() has been called to set logging level to at least the
+  /// corresponding value.
   enum Severity {
-    kAssertFailed = -3,
-    kError = -2,
-    kWarning = -1,
-    kInfo = 0,
+    kAssertFailed = -3,  //!< Assertion failure. abort() will be called.
+    kError = -2,    //!< Fatal error. KaldiFatalError will be thrown.
+    kWarning = -1,  //!< Indicates a recoverable but abnormal condition.
+    kInfo = 0,      //!< Informational message.
   };
-  // An 'enum Severity' value, or a positive number indicating verbosity level.
-  int severity;
-  const char *func;
-  const char *file;
-  int32 line;
+  int severity;      //!< A Severity value, or positive verbosity level.
+  const char *func;  //!< Name of the function invoking the logging.
+  const char *file;  //!< Source file name with up to 1 leading directory.
+  int32 line;        //<! Line number in the source file.
 };
 
-// Class MessageLogger is invoked from the KALDI_ASSERT, KALDI_ERR, KALDI_WARN and
-// KALDI_LOG macros. It formats the message, then either prints it to stderr or
-// passes to the log custom handler if provided, then, in case of the error,
-// throws an std::runtime_exception, in case of failed KALDI_ASSERT calls abort().
-//
-// Note: we avoid using std::cerr for thread safety issues.
-// fprintf(stderr,...) is guaranteed thread-safe, and outputs
-// its formatted string atomically.
+/// Kaldi fatal runtime error exception. This exception is thrown from any use
+/// of the KALDI_ERR logging macro after the logging function, either set by
+/// SetLogHandler(), or the Kaldi's internal one, has returned.
+class KaldiFatalError : public std::runtime_error {
+ public:
+  explicit KaldiFatalError(const std::string &message)
+      : std::runtime_error(message) { }
+  explicit KaldiFatalError(const char *message)
+      : std::runtime_error(message) { }
+
+  /// Returns the exception name, "kaldi::KaldiFatalError".
+  virtual const char *what() const noexcept override {
+    return "kaldi::KaldiFatalError";
+  }
+
+  /// Returns the Kaldi error message logged by KALDI_ERR.
+  const char *KaldiMessage() const { return std::runtime_error::what(); }
+};
+
+// Class MessageLogger is the workhorse behind the KALDI_ASSERT, KALDI_ERR,
+// KALDI_WARN, KALDI_LOG and KALDI_VLOG macros. It formats the message, then
+// either prints it to stderr or passes to the custom logging handler if
+// provided. Then, in case of the error, throws a KaldiFatalError exception, or
+// in case of failed KALDI_ASSERT, calls std::abort().
 class MessageLogger {
-public:
-  /// Constructor stores the info,
+ public:
+  /// The constructor stores the message's "envelope", a set of data which
```

Comment:
Since you cleaned up the destructor output, it would seem to make sense to make this a static function as to not re-construct the class on every log message

Reply:
I'm sorry, I am a little bit dense today. The thing is constructed on the stack, and takes whopping 16 bytes of it. Why should I store a 16-byte buffer in TLS? Or am I totally missing your idea?

Reply:
it's not a bunch of memory.  But it seems like a waste to construct the class and destruct it.  Wasted cpu time

Reply:
Ah, I thought you are talking about the envelope, and you are talking about the whole MessageLoger thing. The class just contains an ostrstream. The logger itself lives on the stack, together with the stream (which in turn just encapsulates an std::string in the end). The stream allocates memory, of course.  But we still need to log to *something*, so this is unavoidable.

It could have been more efficient if we did not pull the string from this logger to append to another stream. But compared to other computational stuff which is going on, it hardly worth an optimization. Verbose logging is for debugging, and normal logging is pretty insignificant.

Reply:
```
  if (GetVerboseLevel() >= 3) {
00007FF614FFA88D  xor         ebx,ebx  
00007FF614FFA88F  lea         r12,[string "kaldi::nnet3::Optimize" (07FF61524E4A0h)]  
00007FF614FFA896  lea         r13,[string "ext\\kaldi\\@"... (07FF61524E6A0h)]  
00007FF614FFA89D  cmp         dword ptr [kaldi::g_kaldi_verbose_level (07FF61531D448h)],3  
00007FF614FFA8A4  jl          kaldi::nnet3::Optimize+0E2h (07FF614FFA932h)  
    CheckComputation(nnet, *computation, true);
00007FF614FFA8AA  mov         r8b,1  
00007FF614FFA8AD  mov         rdx,r9  
00007FF614FFA8B0  mov         rcx,r15  
00007FF614FFA8B3  call        kaldi::nnet3::CheckComputation (07FF614F88DA0h)  
    KALDI_LOG << "Before optimization, max memory use (bytes) = "
00007FF614FFA8B8  lea         rcx,[rbp-50h]  
00007FF614FFA8BC  call        std::basic_ostringstream<char,std::char_traits<char>,std::allocator<char> >::basic_ostringstream<char,std::char_traits<char>,std::allocator<char> > (07FF614E29680h)  
00007FF614FFA8C1  mov         dword ptr [rbp-70h],ebx  
00007FF614FFA8C4  mov         qword ptr [rbp-68h],r12  
00007FF614FFA8C8  lea         edx,[rbx+2Fh]  
00007FF614FFA8CB  mov         rcx,r13 
```
So this is all it boils down to. An ostringstream constructor call.